sw/device/lib/dif/dif_rv_timer.h

```diff
@@ -0,0 +1,321 @@
+ * clock cycle) and a tick step of 20 (since 20 * 50kHz = 1MHz).
+ *
+ * The return value of this function is only an approximation, and the
+ * actual counter frequency ultimately depends on the accuracy of the
+ * clock. The function will return an error if it cannot produce an acceptably
+ * accurate counter frequency using the given clock resolution.
+ *
+ * @param clock_freq The device clock frequency, in Hertz.
+ * @param counter_freq The desired counter frequency, in Hertz.
+ * @param out Tick parameters that will approximately produce the desired
+ *         counter frequency.
+ * @return The result of the operation.
+ */
+dif_rv_timer_approximate_tick_params_result_t
+dif_rv_timer_approximate_tick_params(uint64_t clock_freq, uint64_t counter_freq,
+                                     dif_rv_timer_tick_params_t *out);
+/**
+ * Configuration for initializing the RISC-V timer device.
+ */
+typedef struct dif_rv_timer_config {
+  /**
+   * The number of harts that this device supports time counters for.
+   * Must be at least one.
+   *
+   * This value is determined entirely by the hardware configuration.
+   */
+  uint32_t hart_count;
+
+  /**
+   * The number of comparators (i.e, timers that can be set) associated
+   * with each hart. Must be at least one.
+   *
+   * This value is determined entirely by the hardware configuration.
+   */
+  uint32_t comparator_count;
+} dif_rv_timer_config_t;
+
+/**
+ * State for a RISC-V timer, associated with a particular hardware thread.
+ *
+ * Its member variables should be considered private, and are only provided so
+ * that callers can allocate it.
+ */
+typedef struct dif_rv_timer {
+  mmio_region_t base_addr;
+  dif_rv_timer_config_t config;
+} dif_rv_timer_t;
+
+/**
+ * Initialize a RISC-V timer device with the given configuration.
+ *
+ * This function will deactivate all counters and reset all timers, which should
+ * each be configured and turned on manually after this function returns.
+ *
+ * @param base_addr MMIO region for the device hardware registers.
+ * @param config Configuration for initializing a particular timer.
+ * @param timer_out Out param for the timer device.
+ * @return The result of the operation.
+ */
+dif_rv_timer_result_t dif_rv_timer_init(mmio_region_t base_addr,
+                                        dif_rv_timer_config_t config,
+                                        dif_rv_timer_t *timer_out);
+
+/**
+ * Completely resets a timer device, disabling all IRQs, counters, and
+ * comparators.
+ *
+ * @param timer A timer device.
+ * @return The result of the operation.
+ */
+dif_rv_timer_result_t dif_rv_timer_reset(const dif_rv_timer_t *timer);
+
+/**
+ * Configures the tick params for a particular hart's counter.
+ *
+ * This function should not be called when `hart_id`'s counter is enabled.
+ * The function `dif_rv_timer_approximate_tick_params()` can be used to generate
+ * tick parameter values.
```

Comment:
I think that this kind of state tracking is what we would call a "driver-level concern". Think of it this way:

There are two ways to enforce this constraint: either by tracking an is_enabled bit in software (fast, but requires keeping state), or by pinging the relevant control register (slower, but doesn't require state). In general, we've decided that both of these alternatives are inappropriate for a DIF: we don't want state machines in DIF structures, and we don't want DIFs doing MMIO operations that are not strictly required, since a driver using the DIF might want to go with the fast SW-based solution.

I've added some language to make it clear that we don't check this precondition, though.